nessie/jobs/import_canvas_gradebook_api.py

```diff
@@ -0,0 +1,109 @@
+        app.logger.info(f'Starting Canvas gradebook API import job for term {term_id}, {len(canvas_course_ids)} course sites...')
+
+        rows = []
+        success_count = 0
+        failure_count = 0
+        index = 1
+        for course_id in canvas_course_ids:
+            app.logger.info(f'Fetching Canvas gradebook history for course id {course_id}, term {term_id} ({index} of {len(canvas_course_ids)})')
+            feed = canvas_api.get_course_gradebook_history(course_id)
+            if feed:
+                success_count += 1
+                for submission_version in feed:
+                    user_id = submission_version.get('user_id')
+                    assignment_id = submission_version.get('assignment_id')
+                    submission_id = submission_version.get('id')
+                    graded_at = submission_version.get('graded_at')
+                    rows.append(
+                        encoded_tsv_row(
+                            [
+                                course_id,
+                                term_id,
+                                user_id,
+                                assignment_id,
+                                submission_id,
+                                graded_at,
+                                json.dumps(submission_version),
+                            ],
+                        ),
+                    )
+            else:
+                failure_count += 1
+                app.logger.error(f'Canvas gradebook history import failed for course id {course_id}.')
+            index += 1
+
+        s3_key = f'{get_s3_canvas_api_daily_path()}/gradebook_history_{term_id}.tsv'
+        app.logger.info(f'Will stash {success_count} feeds in S3: {s3_key}')
+        if not s3.upload_tsv_rows(rows, s3_key):
+            raise BackgroundJobError('Error on S3 upload: aborting job.')
+
+        app.logger.info('Will copy S3 feeds into Redshift...')
+        query = resolve_sql_template_string(
+            """
+            DELETE FROM {redshift_schema_canvas_api}_staging.gradebook_history WHERE term_id = '{term_id}';
+            COPY {redshift_schema_canvas_api}_staging.gradebook_history
+                FROM '{loch_s3_canvas_api_data_path_today}/gradebook_history_{term_id}.tsv'
+                IAM_ROLE '{redshift_iam_role}'
+                DELIMITER '\\t'
+                TIMEFORMAT 'YYYY-MM-DDTHH:MI:SSZ';
+            DELETE FROM {redshift_schema_canvas_api}.gradebook_history
+                WHERE term_id = '{term_id}'
+                AND course_id IN
+                (SELECT course_id FROM {redshift_schema_canvas_api}_staging.gradebook_history WHERE term_id = '{term_id}');
+            INSERT INTO {redshift_schema_canvas_api}.gradebook_history
+                (SELECT * FROM {redshift_schema_canvas_api}_staging.gradebook_history WHERE term_id = '{term_id}');
+            DELETE FROM {redshift_schema_canvas_api}_staging.gradebook_history
+                WHERE term_id = '{term_id}';
```

Comment:
@lyttam It might be best to do read the files as an external table in the next iteration. Internal tables will be a memory hog and we just have 1 node on the Redshift cluster. With external tables, we will be able to engage the spectrum layer which gives better performance.